File: social_graph.py
"""Directed social graph with preferential attachment."""

from __future__ import annotations
import random
from models import AgentProfile


class SocialGraph:
    def __init__(self, agents: list[AgentProfile]):
        self.agents = agents
        self.agent_map: dict[str, AgentProfile] = {a.id: a for a in agents}
        self.edges: list[tuple[str, str]] = []

    def build(self):
        """Wire edges based on influence score + shared interests."""
        for a in self.agents:
            a.followers = []
            a.following = []

        for a in self.agents:
            for b in self.agents:
                if a.id == b.id:
                    continue
                shared = len(set(a.interests) & set(b.interests))
                prob = 0.05 + (b.influence_score * 0.3) + (shared * 0.15)
                if random.random() < min(prob, 0.7):
                    a.following.append(b.id)
                    b.followers.append(a.id)
                    self.edges.append((a.id, b.id))

    @property
    def edge_count(self) -> int:
        return len(self.edges)

    def get_influencers(self, n: int = 5) -> list[AgentProfile]:
        return sorted(self.agents, key=lambda a: len(a.followers), reverse=True)[:n]

    def get_followers_of(self, agent_id: str) -> list[str]:
        agent = self.agent_map.get(agent_id)
        return agent.followers if agent else []

    def get_following(self, agent_id: str) -> list[str]:
        agent = self.agent_map.get(agent_id)
        return agent.following if agent else []

    def to_graph_data(self) -> dict:
        """Return nodes and edges for frontend visualization."""
        nodes = [{"id": a.id, "name": a.name, "influence": a.influence_score,
                   "mbti": a.mbti.value, "mood": a.mood.value,
                   "followers": len(a.followers)} for a in self.agents]
        edges = [{"source": s, "target": t} for s, t in self.edges]
        return {"nodes": nodes, "edges": edges}

File: src/main/kotlin/com/resonance/SocialGraph.kt
package com.resonance

import kotlin.random.Random

// ==========================================================================
// Social Graph — pure Kotlin, no external graph lib needed
// ==========================================================================

/**
 * Directed social graph. Nodes are agent IDs, edges represent "follows".
 * Built with preferential attachment weighted by influence score.
 */
class SocialGraph(val agents: List<AgentProfile>) {

    private val adjacency: MutableMap<String, MutableSet<String>> = mutableMapOf()
    private val agentMap: Map<String, AgentProfile> = agents.associateBy { it.id }

    val edgeCount: Int get() = adjacency.values.sumOf { it.size }

    init {
        for (a in agents) {
            adjacency[a.id] = mutableSetOf()
        }
    }

    /**
     * Wire up the graph. Each agent decides who to follow based on:
     * - Target's influence score (higher = more likely to be followed)
     * - Shared interests (boost probability)
     * - Base edge probability
     */
    fun build(baseProbability: Double = 0.15) {
        for (a in agents) {
            for (b in agents) {
                if (a.id == b.id) continue

                var p = baseProbability * b.influenceScore

                // Shared interest boost
                val shared = a.interests.intersect(b.interests.toSet())
                p += 0.05 * shared.size
                p = p.coerceAtMost(0.8)

                if (Random.nextDouble() < p) {
                    adjacency[a.id]!!.add(b.id)
                    a.following.add(b.id)
                    b.followers.add(a.id)
                }
            }
        }
    }

    /** Get all agents that [agentId] follows. */
    fun getFollowing(agentId: String): Set<String> = adjacency[agentId] ?: emptySet()

    /** Get top N agents by follower count. */
    fun getInfluencers(topN: Int = 5): List<AgentProfile> =
        agents.sortedByDescending { it.followers.size }.take(topN)

    /** Get agents who follow anyone in [agentIds]. */
    fun getFollowersOf(agentIds: Set<String>): Set<String> {
        val result = mutableSetOf<String>()
        for (agent in agents) {
            if (agent.following.any { it in agentIds }) {
                result.add(agent.id)
            }
        }
        return result
    }
}
